anton_gadzikovskiy/graph_generator.cpp

```diff
@@ -107,46 +206,168 @@ string print_graph(const Graph& graph) {
-    graph.add_vertex();
-  }
-
-  graph.add_edge(0, 1);
-  graph.add_edge(0, 2);
-  graph.add_edge(0, 3);
-  graph.add_edge(1, 4);
-  graph.add_edge(1, 5);
-  graph.add_edge(1, 6);
-  graph.add_edge(2, 7);
-  graph.add_edge(2, 8);
-  graph.add_edge(3, 9);
-  graph.add_edge(4, 10);
-  graph.add_edge(5, 10);
-  graph.add_edge(6, 10);
-  graph.add_edge(7, 11);
-  graph.add_edge(8, 11);
-  graph.add_edge(9, 12);
-  graph.add_edge(10, 13);
-  graph.add_edge(11, 13);
-  graph.add_edge(12, 13);
-
-  return graph;
+bool is_generated(double prob) {
+  srand(time(0));
+  return (rand() % 100) <= prob;
+}
+
+const VertexId& get_random_vertex_id(const vector<VertexId>& vertices) {
+  srand(time(0));
+  const int random_id = rand() % vertices.size();
+  return vertices[random_id];
+}
+
+void generate_gray_edges(Graph& graph, int depth, int new_vertices_num) {
+  graph.add_vertex();
+  VertexId new_child_vertex_id = 1;
+  for (int i = 0; i < depth; i++) {
+    double step = 100 / depth;
+    double prob = 100 - step * i;
+    if (i < graph.get_vertices_on_depth().size()) {
+      vector<VertexId> temp_vertices_on_depth =
+          graph.get_vertices_on_depth().at(i);
+      for (const auto& from_vertex_id : temp_vertices_on_depth) {
+        for (int k = 0; k < new_vertices_num; k++) {
+          if (is_generated(prob)) {
+            graph.add_vertex();
+            graph.add_edge(from_vertex_id, new_child_vertex_id);
+            new_child_vertex_id++;
+          }
+        }
+      }
+    }
+  }
+}
+
+void generate_green_edges(Graph& graph) {
+  for (const auto& [vertex_id, vertex] : graph.get_vertices()) {
+    if (is_generated(kGreenEdgesProbability)) {
+      graph.add_edge(vertex_id, vertex_id);
+    }
+  }
+}
+
+void generate_yellow_edges(Graph& graph) {
+  const auto& graph_depth = graph.get_vertices_on_depth().size();
+  for (const auto& [depth, vertices] : graph.get_vertices_on_depth()) {
+    if (depth != graph.get_vertices_on_depth().size() - 1) {
+      for (const auto& from_vertex_id :
+           graph.get_vertices_on_depth().at(depth)) {
+        const double prob = graph.get_vertices().at(from_vertex_id).depth *
+                            (100 / (graph_depth - 1));
+        if (is_generated(prob)) {
+          vector<VertexId> vertices_on_deeper_depth;
+          for (const auto& to_vertex_id :
+               graph.get_vertices_on_depth().at(depth + 1)) {
+            if (!graph.is_connected(from_vertex_id, to_vertex_id)) {
+              vertices_on_deeper_depth.push_back(to_vertex_id);
+            }
+          }
+
+          if (!vertices_on_deeper_depth.empty()) {
+            graph.add_edge(from_vertex_id,
+                           get_random_vertex_id(vertices_on_deeper_depth));
+          }
+        }
+      }
+    }
+  }
+}
+
+void generate_red_edges(Graph& graph) {
+  for (const auto& [depth, vertices] : graph.get_vertices_on_depth()) {
+    if (depth < graph.get_vertices_on_depth().size() - 2) {
```

Comment:
Лучше всегда делать ранний выход:
```suggestion
  for (const auto& [depth, vertices] : graph.get_vertices_on_depth()) {
    if (depth >= graph.get_vertices_on_depth().size() - 2) {
      break;
    }
```